### PR TITLE
expose instanceDiscovery flag

### PR DESCRIPTION
**Pull request by siddhijain**
expose instanceDiscovery flag which is set to true by default. 

Reference to python PR https://github.com/AzureAD/microsoft-authentication-library-for-python/commit/dd7b53805efb985f6a85c22f294f20d9adfcc01d

### CODE REVIEW

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java [hunk 1]
```diff
@@ -643,6 +648,18 @@ public T azureRegion(String val) {
             return self();
         }
 
+        /** Historically, MSAL would connect to a central endpoint located at
```
Comment (Avery-Dunn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java:651]: Rather than have a long comment here, it might be better to have a wiki page describing this behavior and make this comment briefer with an aka.ms link to that page.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java [hunk 2]
```diff
@@ -186,4 +188,137 @@ public void aadInstanceDiscoveryTest_AutoDetectRegion_NoRegionDetected() throws
...
+    private static Object[][] getAadClouds(){
+        return new Object[][] {{"https://login.microsoftonline.com/common"} , // #Known to Microsoft
+                {"https://private.cloud/foo"}//Private Cloud
+                };
+    }
+
+    @DataProvider(name = "b2cAdfsClouds")
+    private static Object[][] getNonAadClouds(){
+        return new Object[][] {{"https://contoso.com/adfs"}//ADFS
+//                {"https://login.b2clogin.com/contoso/b2c_policy"},//B2C
```
Comment (Avery-Dunn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Since B2C is commented out, are B2C and ADFS expected to behave the same way?

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java [hunk 3]
```diff
@@ -80,7 +79,16 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
         InstanceDiscoveryMetadataEntry result = cache.get(host);
 
         if (result == null) {
-            doInstanceDiscoveryAndCache(authorityUrl, validateAuthority, msalRequest, serviceBundle);
+            if(msalRequest.application().instanceDiscovery()){
+                doInstanceDiscoveryAndCache(authorityUrl, validateAuthority, msalRequest, serviceBundle);
+            } else {
+                // instanceDiscovery flag is set to False. Do not perform instanceDiscovery.
```
Comment (Avery-Dunn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java:86]: Very minor, but missing a 't': 'Do no**t** perform instanceDiscovery'

File: msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/TestConstants.java [hunk 4]
```diff
@@ -45,7 +45,7 @@ public class TestConstants {
     public final static String B2C_AUTHORITY_SIGN_IN = B2C_AUTHORITY + B2C_SIGN_IN_POLICY;
     public final static String B2C_AUTHORITY_ROPC = B2C_AUTHORITY + B2C_ROPC_POLICY;
     public final static String B2C_READ_SCOPE = "https://msidlabb2c.onmicrosoft.com/msidlabb2capi/read";
-    public final static String B2C_MICROSOFTLOGIN_AUTHORITY = "https://login.microsoftonline.com/tfp/msidlabb2c.onmicrosoft.com/";
+    public final static String B2C_MICROSOFTLOGIN_AUTHORITY = "https://msidlabb2c.b2clogin.com/tfp/msidlabb2c.onmicrosoft.com/";
```
Comment (bgavrilMS) [inline msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/TestConstants.java]: Would be good to have this in a separate PR to help reviewers. I think it's a separate commit, so possibly a rebase error

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java [hunk 5]
```diff
@@ -18,22 +18,22 @@
...
-    private final static String INSTANCE_DISCOVERY_REQUEST_PARAMETERS_TEMPLATE = "?api-version=1.1&authorization_endpoint={authorizeEndpoint}";
-    private final static String HOST_TEMPLATE_WITH_REGION = "{region}.r.{host}";
-    private final static String SOVEREIGN_HOST_TEMPLATE_WITH_REGION = "{region}.{host}";
-    private final static String REGION_NAME = "REGION_NAME";
-    private final static int PORT_NOT_SET = -1;
+    private static final String DEFAULT_TRUSTED_HOST = "login.microsoftonline.com";
+    private static final String AUTHORIZE_ENDPOINT_TEMPLATE = "https://{host}/{tenant}/oauth2/v2.0/authorize";
+    private static final String INSTANCE_DISCOVERY_ENDPOINT_TEMPLATE = "https://{host}:{port}/common/discovery/instance";
+    private static final String INSTANCE_DISCOVERY_REQUEST_PARAMETERS_TEMPLATE = "?api-version=1.1&authorization_endpoint={authorizeEndpoint}";
+    private static final String HOST_TEMPLATE_WITH_REGION = "{region}.r.{host}";
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java]: nit: for the other issue - this will be `{region}.login.microsoft.com`

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java [hunk 6]
```diff
@@ -67,10 +67,9 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
 
             //If region autodetection is enabled and a specific region not already set,
             // set the application's region to the discovered region so that future requests can skip the IMDS endpoint call
-            if (msalRequest.application().azureRegion() == null && msalRequest.application().autoDetectRegion()) {
-                if (detectedRegion != null) {
+            if (msalRequest.application().azureRegion() == null && msalRequest.application().autoDetectRegion()
+                        && null != detectedRegion) {
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java:72]: nit: keep constant on the right side? https://stackoverflow.com/questions/2369226/object-null-or-null-object :)

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java [hunk 7]
```diff
@@ -80,7 +79,16 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
         InstanceDiscoveryMetadataEntry result = cache.get(host);
 
         if (result == null) {
-            doInstanceDiscoveryAndCache(authorityUrl, validateAuthority, msalRequest, serviceBundle);
+            if(msalRequest.application().instanceDiscovery()){
+                doInstanceDiscoveryAndCache(authorityUrl, validateAuthority, msalRequest, serviceBundle);
+            } else {
+                // instanceDiscovery flag is set to False. Do not perform instanceDiscovery.
+                cache.putIfAbsent(host, InstanceDiscoveryMetadataEntry.builder().
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java]: This cache is static / shared by all apps. An app which does not use this flag will still skip instance discovery because it will pull the data from this static cache. Probably a very obscure scenario, but still - you can just return the entry and not put it in the cache.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java [hunk 8]
```diff
@@ -186,4 +188,137 @@ public void aadInstanceDiscoveryTest_AutoDetectRegion_NoRegionDetected() throws
         Assert.assertTrue(entry.aliases().contains("login.microsoft.com"));
         Assert.assertTrue(entry.aliases().contains("sts.windows.net"));
     }
+
+    @DataProvider(name = "aadClouds")
+    private static Object[][] getAadClouds(){
+        return new Object[][] {{"https://login.microsoftonline.com/common"} , // #Known to Microsoft
+                {"https://private.cloud/foo"}//Private Cloud
```
Comment (bgavrilMS) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Consider adding some of the sovereign clouds here

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java [hunk 9]
```diff
@@ -186,4 +188,137 @@ public void aadInstanceDiscoveryTest_AutoDetectRegion_NoRegionDetected() throws
...
+//                {"https://login.b2clogin.com/contoso/b2c_policy"},//B2C
+                };
+    }
+
+    /**
+     * when instance_discovery flag is set to true (by default), an instance_discovery is performed for authorityType = AAD and
+     *     hence, an exception is thrown while making a call to getMetaDataEntry() if instanceDiscoveryResponse is not mocked.
+    */
+    @Test(  dataProvider = "aadClouds",
+            expectedExceptions = StringIndexOutOfBoundsException.class)
```
Comment (bgavrilMS) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: This is a pretty general-purpose exception, which can be thrown from anywhere. It can potentially hide a bug and it makes the test difficult to read / understand. If `StringIndexOutOfBoundsException` comes from HTTP mock, consider replacing this with a strongly typed exception, e.g. HttpMockOutOfRange or smth.

That said, I think there are already tons of tests for this scenario. MSAL tests work by mocking the HTTP layer, so you add fake (request, response) pairs in a queue of mocked requests. If the mock is not consumed, test should fail.

So better test would be:

1. Add HTTP mock for /token but not for instance discovery and ensure instance discoevry cache is empty (i would imagine this already happens)
2. Acquire a token
3. Assert that HTTP mock queue is empty

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java [hunk 10]
```diff
@@ -186,4 +188,137 @@ public void aadInstanceDiscoveryTest_AutoDetectRegion_NoRegionDetected() throws
...
+                .scopes(Collections.singleton("scope")).build();
+
+        MsalRequest msalRequest = new AuthorizationCodeRequest(
+                parameters,
+                app,
+                new RequestContext(app, PublicApi.ACQUIRE_TOKEN_BY_AUTHORIZATION_CODE, parameters));
+
+        URL authorityURL = new URL(authority);
+
+       AadInstanceDiscoveryProvider.getMetadataEntry(
```
Comment (bgavrilMS) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: IMO, it's better to test the end to end rather than this object. So in this case, actually make an `AcquireTokenByAuthCode` and assert that HTTP instance discoevry call is made.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java [hunk 11]
```diff
@@ -186,4 +188,137 @@ public void aadInstanceDiscoveryTest_AutoDetectRegion_NoRegionDetected() throws
...
+                .scopes(Collections.singleton("scope")).build();
+
+        MsalRequest msalRequest = new AuthorizationCodeRequest(
+                parameters,
+                app,
+                new RequestContext(app, PublicApi.ACQUIRE_TOKEN_BY_AUTHORIZATION_CODE, parameters));
+
+        URL authorityURL = new URL(authority);
+
+        AadInstanceDiscoveryProvider.getMetadataEntry(
```
Comment (bgavrilMS) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Same as above, better to test the E2E.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
don't add the new entry to cache, because cache is static
tests can be more useful if they test the end to end experience and mock the http layer

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java (hunk 1 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java:651]: sure, let's talk more about this.

File: msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/TestConstants.java (hunk 4 above)
Comment (siddhijain) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/TestConstants.java]: Yes that would have been better. The second PR took all the changes from the first one as well. Sorry about that.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java (hunk 6 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java:72]: yes that's a good point. I noticed this recently and this needs to be changed at multiple places in the library. I will do this one to start with.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java (hunk 7 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java]: That's a valid point. I will make the changes.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 8 above)
Comment (Avery-Dunn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Would the instance discovery behavior for sovereign clouds be different in some way? Is it worth having both global and sovereign cloud tests here?

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java (hunk 3 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java:86]: Do you not see 't'. I see it :D

Comment (Avery-Dunn) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java:86]: Oops, I meant to put this comment in the other PR. The 't' is missing from the word 'not' in that version: https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/570/files#r1035238440

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 8 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Well, I don't know your codebase, but there may special rules related to public cloud vs sovereign cloud. For example, ESTS-R feature handles these 2 differently and it is also in this general area of the code. So when thinking about different clouds, it's good to have a test case for a sovereign cloud.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java (hunk 1 above)
Comment (rayluo) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java:651]: FWIW, I use one-or-two-paragraphs long comments all the time, because in-place comment gives better locality to next maintainer (or a future you). And it does not rely on an external link thus insusceptible to future bad-link problem (which can happen to aka.ms links, too). Of course, the long comment itself can end with a link to a long long specs or Microsoft Docs link, it is OK for those links to go stale, because the gist is already captured inside the comment.

Besides, will this long comment be automatically used to generate the API docs? If so, then it  is even preferable to write long description here.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 2 above)
Comment (rayluo) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: In this Instance Discovery context, neither B2C nor ADFS provides Instance Discovery endpoint.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java (hunk 1 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java:651]: Yes for the information presented here, I would prefer having these lines here instead of an aka.ms link. If the information is a page long, then it makes sense to move this to another link.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 2 above)
Comment (siddhijain) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: That's right. Instance discovery should not be done for B2C or ADFS even if instance_discovery flag is true.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java (hunk 7 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java]: Made changes in the latest commit.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 9 above)
Comment (siddhijain) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Added end-to-end tests in the latest commit.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java (hunk 11 above)
Comment (siddhijain) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryTest.java]: Added E2E tests in the latest commit.

**Comment by siddhijain** [comment]
> don't add the new entry to cache, because cache is static tests can be more useful if they test the end to end experience and mock the http layer
Done.

**Comment by siddhijain** [comment]
https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/481

**Review by bgavrilMS — APPROVED** [review]

**Review by Avery-Dunn — APPROVED** [review]

**Comment by siddhijain** [comment]
FYI - This branch has been published as a beta release for Azure SDK team to test.

File: msal4j-sdk/README.md [hunk 12]
```diff
@@ -16,7 +16,7 @@ Quick links:
 The library supports the following Java environments:
 - Java 8 (or higher)
 
-Current version - 1.13.3
+Current version - 1.0.0-beta.1
```
Comment (rayluo) [inline msal4j-sdk/README.md]: I do not know how the versioning work in Java. But if its dependency system follows the [precedence behavior of Semantic Versioning](https://semver.org/#spec-item-11):

> Example: ... < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0 < ... < 1.13.3"

then a customer who already making a dependency of some recent version of MSAL Java (for example, ">1.12"), they will not pick up our "1.0.0-beta.1". The general practice is to ship the latest experimental behavior in "1.14.0-beta.1".

In other words, stable version and beta versions all share the same versioning naming space, despite their different suffix.

**Review by rayluo — APPROVED** [review]
Overall LGTM. Just added a minor comment. Still approving.

File: msal4j-sdk/README.md (hunk 12 above)
Comment (siddhijain) [reply msal4j-sdk/README.md]: Yes, since this was our first beta release, we did not know about the nomenclature. I learned about that after publishing the beta release. will keep this in mind next time. Thanks for reviewing!

Comment (g2vinay) [reply msal4j-sdk/README.md]: Yeah, if you intend to release this as beta feature, then it will follow 1.14.0-beta.1, as last stable release is in 1.13.x series, as of this comment.

Comment (siddhijain) [reply msal4j-sdk/README.md]: @g2vinay Do you need this change to happen for testing this feature?

Comment (g2vinay) [reply msal4j-sdk/README.md]: @siddhijain 
the version 1.0.0-beta.1 indicates to user that the beta features will GA in 1.0.0 release, plus v1.0.0-beta.1 is already released, we cannot release it again.
For next beta release, I feel it is a good opportunity to correct the beta versioning and release it as v1.14.0-beta.1 and when the feature is ready for stable release, it can be released in v1.14.0

File: msal4j-sdk/pom.xml [hunk 13]
```diff
@@ -3,7 +3,7 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>com.microsoft.azure</groupId>
 	<artifactId>msal4j</artifactId>
-	<version>1.13.3</version>
+	<version>1.0.0-beta.1</version>
```
Comment (g2vinay) [inline msal4j-sdk/pom.xml]: This version is already released.
why is it marked to an already released version ?

Comment (siddhijain) [reply msal4j-sdk/pom.xml]: Talked offline about release versioning. Resolving it here.
